debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java

```diff
@@ -396,33 +413,37 @@ private Timer getTableScanLogTimer() {
     /**
      * Returns a {@link ChangeRecordEmitter} producing the change records for the given table row.
      */
-    protected abstract ChangeRecordEmitter getChangeRecordEmitter(SnapshotContext snapshotContext, TableId tableId, Object[] row);
+    protected ChangeRecordEmitter getChangeRecordEmitter(SnapshotContext snapshotContext, TableId tableId, Object[] row) {
+        snapshotContext.offset.event(tableId, Instant.ofEpochMilli(getClock().currentTimeInMillis()));
+        return new SnapshotChangeRecordEmitter(snapshotContext.offset, row, getClock());
+    }
 
     /**
      * Returns a valid query string for the specified table, either given by the user via snapshot select overrides or
      * defaulting to a statement provided by the DB-specific change event source.
      *
      * @param tableId the table to generate a query for
-     * @return a valid query string
+     * @return a valid query string or empty if table will not be snapshotted
      */
-    private String determineSnapshotSelect(SnapshotContext snapshotContext, TableId tableId) {
+    private Optional<String> determineSnapshotSelect(SnapshotContext snapshotContext, TableId tableId) {
```

Comment:
Hey @addisonj, this one raised my curiousity -- it's caused by the fact that `Snapshotter` considers snapshot selects as optional. I'm wondering why that is, I'd have said there should always be a snapshot select (filtering of tables is a separate concern of course). So could you shed some light on the reasoning behind this? Thanks!

Reply:
I think the idea was there might be some tables that you want to stream but you wouldn't want to be including in the initial snapshot for some reason, we have some use cases where we have higher volume tables where old data doesn't matter, so we would prefer not to snapshot it. Obviously, we could just put a query in that would `limit 1` or something so I don't think it would be a huge deal to make it be required.

Reply:
Ok, thanks for clarifying, @addisonj. We'll leave it as optional for the time being, may re-consider when merging the current PG-specific `Snapshotter` SPI with the more generic `SnapshotChangeEventSource` contract.